Automattic/simplenote-android
# Fix the issue of not showing the widget configuration screen when applock is enabled

### Issue description
As described in the issue #725, the widget configuration screen is not displayed if the applock is enabled, and the app was already open.
The cause is that the `DefaultAppLock` of the `Passcode-Lock` library starts the `PasscodeUnlockActivity` using the application context, and the flag: `FLAG_ACTIVITY_NEW_TASK`: https://github.com/wordpress-mobile/PasscodeLock-Android/blob/804db0580e7719a36784b8a7c9751614808246ab/library/src/org/wordpress/passcodelock/DefaultAppLock.java#L53-L55. But as we don't customize the [taskAffinity](https://developer.android.com/guide/components/activities/tasks-and-back-stack#Affinities), if the app has an existing task, it will be brought to foreground, and the `PasscodeUnlockActivity` will be added to it, for example:
`Running activities (most recent first):
      TaskRecord{36c9e37 #884 A=com.automattic.simplenote.debug U=0 StackId=382 sz=3}
        Run #2: ActivityRecord{daf4dd u0 com.automattic.simplenote.debug/org.wordpress.passcodelock.PasscodeUnlockActivity t884}
        Run #1: ActivityRecord{ce9c436 u0 com.automattic.simplenote.debug/com.automattic.simplenote.PreferencesActivity t884}
        Run #0: ActivityRecord{33d7dcc u0 com.automattic.simplenote.debug/com.automattic.simplenote.NotesActivity t884}`

But as the widget configuration activity is started by the launcher, it will be added to its task, and not the app's task, so when we finish `PasscodeUnlockActivity` we will go back to the previous screen in the app's task, and not the widget configuration activity.

To solve this there is two ways:
1. The solution suggested in this PR, call `startActivity` from the resumed activity, and not the applicationContext, and don't provider `FLAG_ACTIVITY_NEW_TASK`, this will give us the following task when trying to add the widget:
` Running activities (most recent first):
      TaskRecord{39ed39c #891 A=com.google.android.apps.nexuslauncher U=0 StackId=389 sz=3}
        Run #2: ActivityRecord{a27d305 u0 com.automattic.simplenote.debug/org.wordpress.passcodelock.PasscodeUnlockActivity t891}
        Run #1: ActivityRecord{1d736e4 u0 com.automattic.simplenote.debug/com.automattic.simplenote.NoteWidgetDarkConfigureActivity t891}
        Run #0: ActivityRecord{6c89f9b u0 com.google.android.apps.nexuslauncher/com.android.launcher3.proxy.ProxyActivityStarter t891}
`

2. use a custom `taskAffinity` for the `PasscodeUnlockActivity` to make sure starting it will always create a new task, and then when we go back, we will go to the launcher's task, which contains the widget configuration activity. But it will cause the appearance of the default animation of navigating to a different task, which is similar to navigating to another app:
![passcode](https://user-images.githubusercontent.com/1657201/103791814-3197e080-5043-11eb-8fd3-67db4cb22582.gif)

Personally, I prefer solution 1, but it should be implemented in the passcode library, and not as I did here, I wanted your opinion before creating a PR in the library code, as I don't know the history, and if the use of the `applicationContext` was for some reason, wdyt @theck13 ?

### Fix
The solution suggested by this PR is to extend the `DefaultAppLock` and make small adjustments just for the widget configuration activities while keeping the default behavior for the rest.

### Test

1. Enable PIN lock.
2. Open app.
3. Open navigation drawer.
4. Tap Settings item.
5. Notice Settings is shown.
6. Tap home system navigation button.
7. Long-press home screen.
8. Tap Widgets option.
9. Scroll to Simplenote widget.
10. Notice To-Do List preview image.
11. Long-press Note widget.
12. Place Note widget on home screen.
13. Notice PIN lock is shown.
14. Enter passcode.
15. Confirm that the Select Dialog is shown.

### Review
Only one developer is required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.
